File: C++/Problems/Problems/problem_322.cpp
#include "problem_322.h"

bool problem_322::test() {
    std::vector coins { 1, 2, 5 };
    int amount = 11;

    int expected = 3;

    auto result = coinChange(coins, amount);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_322::coinChange(std::vector<int> &coins, int amount) {
    std::vector<int> dp(amount + 1);
    dp[0] = 0;
    for (int i = 1; i <= amount; ++i)
    {
        dp[i] = INT_MAX;
        for (auto c : coins)
        {
            int req = i - c;
            if (req >= 0 && dp[req] != INT_MAX)
                dp[i] = std::min(dp[i], dp[req] + 1);
        }
    }
    return dp[amount] == INT_MAX ? -1 : dp[amount];
}

File: Python/Problems/Problems/problem322.py
import heapq
import math
import queue
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter, deque

from .problem import Problem

class Problem322(Problem):
    def test(self) -> bool:
        coins = [1, 2, 5]
        amount = 11

        expected = 3

        result = self.coinChange(coins, amount)

        print(expected)

        return result == expected

    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0]
        for i in range(1, amount + 1):
            dp.append(sys.maxsize)
            for c in coins:
                req = i - c
                if req >= 0 and dp[req] != sys.maxsize:
                    dp[i] = min(dp[i], dp[req] + 1)

        return -1 if dp[amount] == sys.maxsize else dp[amount]